File: src/leetCode/Ex39.java
package leetCode;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

public class Ex39 {
    private List<List<Integer>> res;
    private Deque<Integer> temp;
    private int t;
    private int[] cs;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        res = new ArrayList<>();
        temp = new LinkedList<>();
        t = target;
        cs = candidates;
        helper(0, 0);
        return res;
    }

    private void helper(int i, int cur){
        if(cur==t){
            res.add(new ArrayList<>(temp));
            return;
        }
        if(i==cs.length){
            return;
        }
        int j;
        for (j = 0; cur+cs[i]*j<=t; j++) {
            if(j>0){
                temp.push(cs[i]);
            }
            helper(i+1, cur+cs[i]*j);
        }
        for (int j2 = 1; j2 < j; j2++) {
            temp.pop();
        }
    }
}

File: src/leetCode/ex39.kt
package leetCode

fun combinationSum(candidates: IntArray, target: Int): List<List<Int>> {
    return dfs(candidates.sorted().toIntArray(),target).distinctBy {
        it.toHash()
    }
}

private fun List<Int>.toHash():Int{
    var res = 1
    forEach { res *=it.hashCode()+3 }
    return res+size
}

private fun dfs(candidates: IntArray,target: Int):MutableList<MutableList<Int>>{
    val res = mutableListOf<MutableList<Int>>()
    candidates.filter { it==target }.forEach { res.add(mutableListOf(it)) }
    candidates.filter { target-it>0 }.forEach { item->
        res.addAll(dfs(candidates,target-item).apply {
            forEach { it.add(item) }
        })
    }
    return res
}

fun main(args: Array<String>) {
    combinationSum(intArrayOf(4,2,8),8).forEach { println(it) }
}